### PR TITLE
Add support for OpenShift 4

### PR DESCRIPTION
**Pull request by ZuGao**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the ./CHANGELOG.md.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [v1.3.0]
+### Changed
+Added openshift 4 support
```
Comment (tobru) [inline CHANGELOG.md:11]:
```suggestion
Added OpenShift 4 support
```

File: class/defaults.yml [hunk 2]
```diff
@@ -1,5 +1,11 @@
 parameters:
   crossplane:
+    distribution: default
```
Comment (simu) [inline class/defaults.yml:3]: Not having `distribution` set to the cluster's distribution fact (`${facts:distribution}`) by default feels counter-intuitive. 

Additionally, there's a high chance that someone will set this parameter to `${facts:distribution}` somewhere causing the component to break in unexpected ways for clusters which have a distribution which isn't listed in the lookup table.

For those reasons, I don't think using lookup tables is the right choice here. Instead, if you want to handle the changes required for OpenShift in the component, I would instead suggest writing a postprocessing filter which removes the `runAsUser` field from the manifests rendered from the Helm chart.

Alternatively, you could add logic to the component to deploy an SCC which allows using UID 66532 and RBAC rules which grant the Crossplane components the right to use the SCC. Of course you'd have to do this only when `${facts:distribution}` is `openshift4`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [v1.3.0]
+### Changed
+Added OpenShift 4 support
```
Comment (simu) [inline CHANGELOG.md:11]: You can add this to v1.2.0, as long as you haven't tagged that version, you don't need to make a new minor version for each change.

**Review by simu — CHANGES_REQUESTED** [review]
See inline comment for my thoughts on the implementation after actually seeing it.

Additionally, if you end up with an implementation which introduces new parameters, please make sure to add documentation for the new parameters in `docs/moules/ROOT/pages/parameters.adoc`.

File: class/defaults.yml (hunk 2 above)
Comment (ZuGao) [reply class/defaults.yml:3]: After careful consideration the post processing approach seemed the best, that's why I move into that direction.

**Comment by ZuGao** [comment]
I removed the changes I have done to Makefile. I left the changes to ModuleSync and will work on ModuleSync's PR to make it green.

File: postprocess/patch_operator_deployment.jsonnet [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                    c.securityContext.runAsUser,
+              },
+            }
+            for c in deployment.spec.template.spec.containers
+          ],
+        },
+      },
+    },
+  },
+}
```
Comment (simu) [inline postprocess/patch_operator_deployment.jsonnet:29]: I would make this more generic, so any occurrences of `runAsUser` are patched, instead of having to keep a list of files to patch here.

Something like:

```suggestion
local helmchart_dir = std.extVar('output_path');
local on_openshift4 = inv.parameters.facts.distribution;

local removeRunAsUser(obj) =
  if obj.kind == 'Deployment' then
    obj {
     spec+: {
      template+: {
        spec+: {
          containers: [
            c {
              securityContext+: {
                runAsUser: null
              },
            }
            for c in deployment.spec.template.spec.containers
          ],
        },
      },
    }
  else
    obj;

if on_openshift4 then
  com.fixupDir(helmchart_dir, removeRunAsUser)
else 
  {}
```

See https://syn.tools/commodore/reference/commodore-libjsonnet.html#_fixupdirdir_fixupfn for details how `com.fixupDir()` works.

**Review by simu — CHANGES_REQUESTED** [review]
As far as I can see, golden tests are not enabled for the component in https://github.com/projectsyn/component-crossplane/blob/master/.sync.yaml (which should be named `.sync.yml` to be picked up by modulesync as noted in the modulesync PR), so modulesync won't configure the golden test make targets or GitHub actions. See https://github.com/projectsyn/component-rook-ceph/blob/master/.sync.yml for an example sync config which enables the same set of golden tests as you have here.

Additionally, I'd use [`com.fixupDir()`](https://syn.tools/commodore/reference/commodore-libjsonnet.html#_fixupdirdir_fixupfn) to make sure no deployments created through the helmchart set `runAsUser` on OpenShift 4 (see inline comment)

File: tests/golden/openshift4/crossplane/crossplane/01_helmchart/crossplane/templates/rbac-manager-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+              cpu: 1000m
+              memory: 512Mi
+            requests:
+              cpu: 500m
+              memory: 256Mi
+          securityContext:
+            allowPrivilegeEscalation: false
+            readOnlyRootFilesystem: true
+            runAsGroup: 65532
+            runAsUser: 65532
```
[the remark this answers is not in the document]
Comment (ZuGao) [reply tests/golden/openshift4/crossplane/crossplane/01_helmchart/crossplane/templates/rbac-manager-deployment.yaml:53]: Fixed with `com.fixupDir()` as suggested above

File: postprocess/patch_operator_deployment.jsonnet (hunk 4 above)
Comment (ZuGao) [reply postprocess/patch_operator_deployment.jsonnet:29]: I implemented almost the same logic while fixing the review :)

File: .sync.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+:global:
```
Comment (simu) [inline .sync.yml:1]: Please put modulesync-managed config in place for this PR (golden tests GitHub actions etc.). Otherwise, it's quite possible that you'll commit stale golden test outputs, and I'm not going to manually verify that the committed golden test outputs match the actual output of compiling the component.

I would just move this config to a separate PR and then make sure you run and merge the sync for this repo. You can then rebase this PR on the latest master branch.

**Review by simu — CHANGES_REQUESTED** [review]

File: .sync.yml (hunk 6 above)
Comment (ZuGao) [reply .sync.yml:1]: I was having doubts as to how to move on with modulesync. Ok I will add that file in a new PR.

Comment (ZuGao) [reply .sync.yml]: Let's try another round of review :), Hopefully this time around everything clicks

File: tests/golden/openshift4/crossplane/crossplane/01_helmchart/crossplane/crds/pkg.crossplane.io_providers.yaml [hunk 7]
```diff
@@ -0,0 +1,331 @@
+null
+---
+apiVersion: apiextensions.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (simu) [reply tests/golden/openshift4/crossplane/crossplane/01_helmchart/crossplane/crds/pkg.crossplane.io_providers.yaml:3]: No ArgoCD doesn't about `null` resources in the YAML stream. The `null` entries are generated because the Helm chart has an empty line above the first document separator for the CRDs, see e.g. https://github.com/crossplane/crossplane/blob/216d8c9b1ee10fe450cde7982eebffacc6724ee8/cluster/charts/crossplane/crds/apiextensions.crossplane.io_compositeresourcedefinitions.yaml#L1-L2

File: tests/defaults.yml [hunk 8]
```diff
@@ -1,7 +1,36 @@
...
     charts:
-      crossplane: 1.7.0
+      crossplane: 1.1.1
     images:
       crossplane:
-        tag: v1.7.0
+        registry: docker.io
+        image: crossplane/crossplane
+        tag: v1.1.1
+        pullPolicy: IfNotPresent
```
[the remark this answers is not in the document]
Comment (simu) [reply tests/defaults.yml]: The component defaults have never been updated to use v1.7, the initial `tests/defaults.yml` and `tests/openshift4.yml` were created with v1.7 instead of just using the component defaults which I didn't catch in that PR.

**Comment by simu** [comment]
> > Also, I think the golden tests output is not correct at the moment (I get an error for `make golden-diff -e instance=openshift4` locally), but the checks don't catch it because of the misconfiguration in the modulesync configuration.
> 
> For some reason the indentation is different in the compiled manifests between `defaults` and `openshift4`. The test manifests are the same between `defaults` and `openshift4` except for the `runAsUser`

Indentation will be changed by postprocessing, running `make gen-golden -e instance=openshift4` generates a diff for me locally with the latest version of this PR, so you probably want to commit those changes.

File: .sync.yml [hunk 9]
```diff
@@ -2,11 +2,15 @@
...
       - defaults
       - openshift4
 
 docs/antora.yml:
   name: crossplane
   title: crossplane
+
+.github/workflows/test.yaml:
+  makeTarget: test -e instance=${{ matrix.instance }}
+  goldenTest_makeTarget: golden-diff -e instance=${{ matrix.instance }}
```
Comment (simu) [inline .sync.yml:16]: Once more, please move this fix to a separate PR (or commit it directly to master) and rerun modulesync to get the updated GitHub actions config.

Comment (ZuGao) [reply .sync.yml:16]: sure lets do that

Comment (ZuGao) [reply .sync.yml:16]: ModuleSync [PR](https://github.com/projectsyn/component-crossplane/pull/21)

**Review by simu — APPROVED** [review]
Implementation LGTM now. Please clean up the PR title and set an appropriate label (probably `enhancement`) so that the change shows up correctly and nicely formatted in the generated changelog for the next release.

**Comment by tobru** [comment]
Why is the default version still 1.1.1? We should have the latest 1.7 as a default version. If this doesn't fit in this PR which I suppose it won't, please do that in another PR.

**Comment by ZuGao** [comment]
> Why is the default version still 1.1.1? We should have the latest 1.7 as a default version. If this doesn't fit in this PR which I suppose it won't, please do that in another PR.

I will get more information from the relevant parties that use this component if that's fine for them to update the default version

**Comment by tobru** [comment]
> I will get more information from the relevant parties that use this component if that's fine for them to update the default version

We should always keep the components up-to-date, as we have versioned components and possibilities to pin versions in the configuration hierarchy.

**Comment by ZuGao** [comment]
@tobru I will open a new PR to update to the latest stable version.
